### PR TITLE
[WIP] more vector defuns and lisp

### PR DESCRIPTION
**Pull request by mjfeller**
Depends on the vectors and buffers work on https://github.com/Wilfred/remacs/pull/202

This is my first WIP PR. I'm looking for advice and a bit of help. 

Not sure why the whitespace got all messed up in C. I'll look into that

### CODE REVIEW

File: rust_src/remacs-sys/lib.rs [hunk 1]
```diff
@@ -33,40 +35,54 @@ pub const CHAR_MODIFIER_MASK: char_bits =
...
+    PVEC_FONT, /* Should be last because it's used for range checking.  */
+}
 
 #[derive(Debug)]
 #[repr(C)]
 pub struct vectorlike_header {
     pub size: libc::ptrdiff_t,
 }
 
+// XXX: this can also be char on some archs
```
Comment (DavidDeSimone) [inline rust_src/remacs-sys/lib.rs]: bits_word is now always a size_t as of https://github.com/Wilfred/remacs/pull/201, we can remove this comment.

**Comment by mjfeller** [comment]
I cleaned up the formatting issues, but there is a bug open on rustfmt right now, https://github.com/rust-lang-nursery/rustfmt/issues/1668, that is causes an issue when formatting

```rust
#[cfg(all(not(test) ), target_os = "macos"))]
```

File: rust_src/remacs-sys/lib.rs (hunk 1 above)
Comment (mjfeller) [inline rust_src/remacs-sys/lib.rs]: Thanks, removed in https://github.com/Wilfred/remacs/pull/206/commits/184365a9d58df32a4a971d350c32a48a08e464be

**Comment by mjfeller** [comment]
It would seem the unintentional changes in the C files are due to emacs replacing tabs with spaces. Should I put the tabs back in?

**Comment by birkenfeld** [comment]
You'll have to use c-set-style with 'gnu for the Emacs C files.

File: rust_src/src/lisp.rs [hunk 2]
```diff
@@ -393,6 +392,90 @@ impl LispObject {
...
+impl LispObject {
+    pub fn is_thread(self) -> bool {
+        match self.as_vectorlike() {
+            Some(vl) => vl.is_pseudovector(PseudovecType::PVEC_THREAD),
+            None => false,
+        }
+    }
+}
+
+impl LispObject {
```
Comment (birkenfeld) [inline rust_src/src/lisp.rs]: These fns should be written like `is_vector`:
```
self.as_vectorlike().map_or(false, |v| v.is_pseudovector(XXX))
```

File: rust_src/src/character.rs [hunk 3]
```diff
@@ -21,3 +21,9 @@ fn max_char() -> LispObject {
 fn characterp(object: LispObject, _ignore: LispObject) -> LispObject {
     LispObject::from_bool(object.is_character())
 }
+
+/// Return t if OBJECT is a character or a string.
+#[lisp_fn(min = "1")]
+fn character_or_string_p(object: LispObject) -> LispObject {
```
Comment (birkenfeld) [inline rust_src/src/character.rs]: This is `char-or-string-p` in Lisp.

File: rust_src/src/lisp.rs [hunk 4]
```diff
@@ -924,3 +960,61 @@ pub fn check_number_coerce_marker(x: LispObject) -> LispObject {
...
+                    } else if self.is_mutex() {
+                        Qmutex
+                    } else if self.is_condition_variable() {
+                        Qcondition_variable
+                    } else {
+                        Qvector
+                    }
+                }
+                LispType::Lisp_Float => Qfloat,
+                _ => panic!(),
```
Comment (birkenfeld) [inline rust_src/src/lisp.rs]: A message would be nice here (including the value of self).

File: rust_src/src/lisp.rs [hunk 5]
```diff
@@ -924,3 +960,61 @@ pub fn check_number_coerce_marker(x: LispObject) -> LispObject {
...
+            };
+            LispObject(tp)
```
Comment (birkenfeld) [inline rust_src/src/lisp.rs]: Using ::from_raw() would be better.

**Comment by birkenfeld** [comment]
The big dependency is now merged. Would you like to you rebase and squash these commits?

**Comment by mjfeller** [comment]
My git-fu failed me and I've really made a mess of this branch. I've created a new branch with my changes from the latest master here. I'm going to close this PR and open a new one.

**Pull request closed without merging**